### DISCUSSION TITLE
Is it possible to add better image alternative text?

### DISCUSSION THREAD

**Discussion by nzyeehaa** [datenstrom/community#480] [Ask a question] [answered]
website users often dont know much about technical aspects of websites, they are point and click wizards. Which is why Yellow is perfect for them.

it annoys me when every image alt tag shows as "Image"... (as it is not good SEO)
this can of course be changed in the Yellow settings file _**ImageAlt: Image**_ but they will still **all be the same** unless a user adds a title/alt manually as shown here...
`[image blue-fish.jpg "saltwater bluefish (Pomatomus saltatrix) also known as tailor / elf / shad"]`

but they often _forget._...

I modified the editor.js file to add "title" when an image is uploaded eg: `[image blue-fish.jpg "title"]` simply to prompt them to add a title.

then i modified image.php to check if a title has been added, if it still shows "_title_"...
replace it with the "**file name**" which will then be **unique**(better for SEO), and probably describe the image much better...
eg: `[image blue-fish.jpg "title"]` the title has not been filled in here so it will now show its "file name" for title/alt...

so in html it now looks like: `<img src="/media/images/blue-fish.jpg" alt="blue-fish.jpg" title="blue-fish.jpg" />`

happy to share code adjustments if anyone wants it, maybe it will find its way into yellow core some time.

a.t.m. you would have to re-enter it every time there was an update to edit.js or image.php

**Comment by nzyeehaa** [comment, marked as answer]
@markseu re: https://github.com/datenstrom/yellow/issues/480#issuecomment-575981577

I dont believe it is worthy of an extension.
it is only a few lines of code

some people may still be happy with **all** their website images being called "**image**",
it just is not my preference....

I currently have a dozen changes I make to some core files each time the site is updated, to suit my preferences and to make it easier for the people I deal with....
_...(the line numbers may have changed due to updates)_

I set image ImageUploadWidthMax: 640 in system.ini  ... 
...then remove the image 'sizes' in image.php...
  45   // if ($width && $height) $output .= " width=\"".htmlspecialchars($width)."\" height=\"".htmlspecialchars($height)."\"";
...and set all image sizes through CSS. (I realise google also doesnt like this much - however this is to make it simple for the end user, and gives much better conformity throughout the site.)

I add {target=_blank} to all links, and users can remove it if its not an external link...
***adjust edit.js***
```
588 [link](url){target=_blank}
```

I set the help screen to open in a blank tab...
***adjust edit.js*** 
```
667 // window.open(this.getText("HelpUrl", "yellow"), "_self");
       window.open(this.getText("HelpUrl", "yellow"), "_blank");
```

and as for the images titles as above...
***adjust images.php ****
```
35  // if (empty($alt)) $alt = $this->yellow->system->get("imageAlt");
    if (empty($alt) || $alt=="title") $alt = $name;
```

***adjust edit.js***
```
840 // after image has uploaded
textNew = "[image "+result.location.substring(yellow.system.coreimageLocation.length)+" \"title\"]";
```

hope that helps someone, and please remember the line numbers may change due to updates

**Comment by markseuffert** [comment]
Thanks for sharing your tips, much appreciated. Here are some thoughts: 

The API has an event system. Instead of changing the source code of the edit extension, you could put your functionality into a new extension. It's definitely more work, but your changes would survive a website update. For example handle Markdown in function `onEditContentFile()`.

The image extension is responsible for creating the HTML code for images. Perhaps it should be more flexible, when there's no "alternative text" in the Markdown. For example a setting `auto` that tells the image extension to find a more suitable text. Just some ideas for the future.
